src/app/core/services/ntp-time-service.ts

```diff
@@ -0,0 +1,52 @@
+import { TimeService } from '../abstractions/time-service'
+import { Injectable } from '@angular/core'
+import { NtpEventObserver } from './models/ntp-event-observer'
+import { NtpEvent } from '../models/ntp-event'
+import { ConnectionStatusObserver } from './connection-status-observer.service'
+
+const NTP_REQUEST_INTERVAL_IN_MS: number = 10_000
+const MAX_NTP_TIME_OFFSET_SAMPLE_SIZE: number = 15
+const NTP_TIME_OFFSET_SAMPLE_SIZE: number = 3
+
+@Injectable({
+  providedIn: 'root',
+})
+export class NtpTimeService implements TimeService {
+  private timeOffsetsInMs: number[] = []
+  private timeDriftInMs: number = 0
+
+  constructor(
+    private readonly ntpEventObserver: NtpEventObserver,
+    private readonly connectionStatusObserver: ConnectionStatusObserver
+  ) {
+    this.ntpEventObserver.subscribeToNtpEvents(ntpEvent => {
+      const timeOffsetInMs: number = this.calculateTimeOffset(ntpEvent)
+      this.updateTimeOffsetsAndTimeDrift(timeOffsetInMs)
+    })
+
+    this.connectionStatusObserver.subscribeToReconnect(() => this.requestNtpSamples(MAX_NTP_TIME_OFFSET_SAMPLE_SIZE))
+
+    setInterval(() => this.requestNtpSamples(NTP_TIME_OFFSET_SAMPLE_SIZE), NTP_REQUEST_INTERVAL_IN_MS)
+    this.requestNtpSamples(MAX_NTP_TIME_OFFSET_SAMPLE_SIZE)
+  }
+
+  private calculateTimeOffset(ntpEvent: NtpEvent): number {
+    const now: number = Date.now()
+    const clockDifferenceInMs: number = (ntpEvent.timestamp - ntpEvent.clientTimestamp + (ntpEvent.timestamp - now)) / 2
+    const roundTripDelayInMs: number = now - ntpEvent.clientTimestamp
+    return clockDifferenceInMs + roundTripDelayInMs / 2
+  }
+
+  private updateTimeOffsetsAndTimeDrift(newTimeOffsetInMs: number): void {
+    this.timeOffsetsInMs = this.timeOffsetsInMs.concat(newTimeOffsetInMs).slice(-MAX_NTP_TIME_OFFSET_SAMPLE_SIZE)
+    this.timeDriftInMs = this.timeOffsetsInMs.reduce((timeDriftSumInMs, timeOffsetInMs) => timeDriftSumInMs + timeOffsetInMs, 0) / Math.max(this.timeOffsetsInMs.length, 1)
+  }
+
+  private requestNtpSamples(sampleSize: number): void {
+    Array.from({ length: sampleSize }).forEach(() => this.ntpEventObserver.emitNtpEvent(Date.now()))
+  }
+
+  public getTime(): number {
```

Comment:
```suggestion
  public getTimeWithDrift(): number {
```

Just a suggestion :)

Reply:
There is nothing in the abstraction that states anything about time drift or NTP, so I will leave it as is. We could make an implementation that calls Date.now() and it would still be a valid implementation - though it won't fit our needs.